File: app/src/main/java/org/andstatus/app/graphics/CachedImage.java
/*
 * Copyright (c) 2016-2017 yvolk (Yuri Volkov), http://yurivolkov.com
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.andstatus.app.graphics;

import android.graphics.Bitmap;
import android.graphics.Point;
import android.graphics.Rect;
import android.graphics.drawable.Drawable;
import android.support.annotation.NonNull;

/**
 * @author yvolk@yurivolkov.com
 */
public class CachedImage {
    public static final Bitmap.Config BITMAP_CONFIG = Bitmap.Config.ARGB_8888;
    public static final Rect EMPTY_RECT = new Rect(0, 0, 0, 0);
    public static final Bitmap EMPTY_BITMAP = newBitmap(1);
    public static final CachedImage EMPTY = new CachedImage(-1, EMPTY_BITMAP, EMPTY_RECT).makeExpired();
    public static final CachedImage BROKEN = new CachedImage(-2, EMPTY_BITMAP, EMPTY_RECT).makeExpired();
    public final long id;
    private final Bitmap bitmap;
    protected final Drawable source;
    private volatile boolean expired = false;

    public CachedImage(long imageId, @NonNull Bitmap bitmap, @NonNull Rect srcRect) {
        id = imageId;
        this.bitmap = bitmap;
        source = new BitmapSubsetDrawable(bitmap, srcRect);
    }

    public CachedImage(long imageId, @NonNull Drawable drawable) {
        id = imageId;
        bitmap = EMPTY_BITMAP;
        source = drawable;
    }

    private static Bitmap newBitmap(int size) {
        return Bitmap.createBitmap(size, size, BITMAP_CONFIG);
    }

    boolean isBitmapRecyclable() {
        return !expired && !EMPTY_BITMAP.equals(bitmap);
    }

    @NonNull
    Bitmap getBitmap() {
        return bitmap;
    }

    public boolean isExpired() {
        return expired;
    }

    CachedImage makeExpired() {
        this.expired = true;
        return this;
    }

    public Drawable getDrawable() {
        return source;
    }

    public Point getImageSize() {
        return new Point(source.getIntrinsicWidth(), source.getIntrinsicHeight());
    }
}

File: app/src/main/kotlin/org/andstatus/app/graphics/CachedImage.kt
/*
 * Copyright (c) 2016-2021 yvolk (Yuri Volkov), http://yurivolkov.com
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package org.andstatus.app.graphics

import android.graphics.Bitmap
import android.graphics.Point
import android.graphics.Rect
import android.graphics.drawable.Drawable
import org.andstatus.app.util.IsEmpty

/**
 * @author yvolk@yurivolkov.com
 */
class CachedImage : IsEmpty {
    val id: Long
    val foreignBitmap: Boolean
    @Volatile
    private var bitmap: Bitmap?
    @Volatile
    private var source: Drawable?
    @Volatile
    private var expired = false

    constructor(imageId: Long, bitmap: Bitmap, srcRect: Rect, foreignBitmap: Boolean) {
        id = imageId
        this.foreignBitmap = foreignBitmap
        this.bitmap = bitmap
        source = BitmapSubsetDrawable(bitmap, srcRect, imageId)
    }

    constructor(imageId: Long, drawable: Drawable) {
        id = imageId
        foreignBitmap = false
        bitmap = EMPTY_BITMAP
        source = drawable
    }

    override val isEmpty: Boolean
        get() = (bitmap == null || bitmap === EMPTY_BITMAP) && source == null

    fun isExpired(): Boolean {
        return expired
    }

    fun makeExpired(): Bitmap? {
        val recycledBitmap = if (nonEmpty) bitmap else null
        if (nonEmpty) {
            expired = true
            bitmap = null
            source = null
        }
        return recycledBitmap
    }

    fun getDrawable(): Drawable {
        return source ?: EMPTY.getDrawable()
    }

    fun getImageSize(): Point {
        return source?.let {
            Point(it.getIntrinsicWidth(), it.getIntrinsicHeight())
        }  ?: Point (0, 0)
    }

    companion object {
        val BITMAP_CONFIG: Bitmap.Config = Bitmap.Config.ARGB_8888
        val EMPTY_RECT: Rect = Rect(0, 0, 0, 0)
        val EMPTY_BITMAP = Bitmap.createBitmap(1, 1, BITMAP_CONFIG)
        val EMPTY = CachedImage(-1, EMPTY_BITMAP, EMPTY_RECT, true)
        val BROKEN = CachedImage(-2, EMPTY_BITMAP, EMPTY_RECT, true)
    }
}
